File: frontend/lib/test-source.ts
const TEST_SOURCE_LABELS = {
  cambridge: "Cambridge Official",
  real_exam: "Recent Exam Papers",
  custom: "Exam Practice Tests",
} as const;

type TestSourceKey = keyof typeof TEST_SOURCE_LABELS;

const LEGACY_SOURCE_LABELS: Record<string, TestSourceKey> = {
  "cambridge official": "cambridge",
  "real exam material": "real_exam",
  "custom practice": "custom",
  "recent exam papers": "real_exam",
  "exam practice tests": "custom",
};

export function getTestSourceKey(source: string | null | undefined): TestSourceKey | null {
  const normalized = String(source ?? "").trim().toLowerCase();
  if (!normalized) {
    return null;
  }
  if (normalized === "cambridge" || normalized.includes("cambridge")) {
    return "cambridge";
  }
  if (normalized === "real_exam" || normalized.includes("real exam") || normalized.includes("recent exam papers")) {
    return "real_exam";
  }
  if (normalized === "custom" || normalized.includes("custom practice") || normalized.includes("exam practice tests")) {
    return "custom";
  }
  return LEGACY_SOURCE_LABELS[normalized] ?? null;
}

export function getTestSourceLabel(source: string | null | undefined): string {
  const key = getTestSourceKey(source);
  return key ? TEST_SOURCE_LABELS[key] : TEST_SOURCE_LABELS.custom;
}

export function getTestSourceDetail(source: string | null | undefined, detail: string | null | undefined): string {
  const normalizedDetail = String(detail ?? "").trim();
  if (!normalizedDetail) {
    return getTestSourceLabel(source);
  }

  const detailKey = getTestSourceKey(normalizedDetail);
  if (detailKey) {
    return TEST_SOURCE_LABELS[detailKey];
  }

  return normalizedDetail;
}

export function matchesTestSourceFilter(
  source: string | null | undefined,
  detail: string | null | undefined,
  requestedSource: string | null | undefined,
): boolean {
  const requestedKey = getTestSourceKey(requestedSource);
  if (!requestedKey) {
    return true;
  }

  return getTestSourceKey(source) === requestedKey || getTestSourceKey(detail) === requestedKey;
}

export const testSourceOptions = [
  { id: "", label: "All Materials" },
  { id: "cambridge", label: TEST_SOURCE_LABELS.cambridge },
  { id: "real_exam", label: TEST_SOURCE_LABELS.real_exam },
  { id: "custom", label: TEST_SOURCE_LABELS.custom },
] as const;

File: backend/app/services/test_source.py
from __future__ import annotations

from app.models.enums import TestSource


TEST_SOURCE_LABELS: dict[TestSource, str] = {
    TestSource.CAMBRIDGE: "Cambridge Official",
    TestSource.REAL_EXAM: "Recent Exam Papers",
    TestSource.CUSTOM: "Exam Practice Tests",
}


def normalize_test_source(source: TestSource | str | None) -> TestSource:
    if isinstance(source, TestSource):
        return source

    normalized = str(source or "").strip().lower()
    if normalized == TestSource.CAMBRIDGE.value or "cambridge" in normalized:
        return TestSource.CAMBRIDGE
    if normalized == TestSource.REAL_EXAM.value or "real exam" in normalized or "recent exam papers" in normalized:
        return TestSource.REAL_EXAM
    return TestSource.CUSTOM


def get_test_source_label(source: TestSource | str | None) -> str:
    return TEST_SOURCE_LABELS[normalize_test_source(source)]


def normalize_test_source_detail(source: TestSource | str | None, detail: str | None) -> str:
    normalized_detail = str(detail or "").strip().lower()
    if normalized_detail in {
        "",
        "cambridge official",
        "real exam material",
        "custom practice",
        "recent exam papers",
        "exam practice tests",
    }:
        return get_test_source_label(source)
    return str(detail or "").strip()
